File: mac/VoiceTests/TextInjectorRoutingTests.swift
import XCTest
@testable import Voice

/// prd-insights-dashboard.md rule 3 / build spec invariant 14: a dictation must never be pasted
/// into Voice itself.
///
/// This became possible the moment Voice grew a dock icon and a window. Before that the app had no
/// focusable window, so "the frontmost app is Voice" could not happen; now it can, and the failure
/// is silent — the transcript would be typed into the Insights window, land nowhere, and be gone
/// before the user understood what they were looking at.
final class TextInjectorRoutingTests: XCTestCase {

    /// The test bundle is hosted inside Voice.app, so `Bundle.main` here is the app itself — the
    /// same value the injector compares against at runtime.
    private var own: String { TextInjector.ownBundleId }

    func testDictationIntoVoiceItselfGoesToTheClipboard() {
        XCTAssertTrue(TextInjector.mustUseClipboard(targetBundleId: own, secureInputActive: false))
    }

    func testOrdinaryAppsStillGetAPaste() {
        for bundleId in ["com.apple.mail", "com.tinyspeck.slackmacgap", "com.apple.Notes", "com.apple.dt.Xcode"] {
            XCTAssertFalse(
                TextInjector.mustUseClipboard(targetBundleId: bundleId, secureInputActive: false),
                "\(bundleId) must still receive a normal paste",
            )
        }
    }

    func testAnUnknownTargetPastesRatherThanBeingTreatedAsVoice() {
        // `FrontmostContext.current()` returns nil when there is no frontmost app, and an app can
        // report a nil bundle id. Reading "I don't know" as "it might be me" would divert ordinary
        // dictations to the clipboard for no reason the user could see.
        XCTAssertFalse(TextInjector.mustUseClipboard(targetBundleId: nil, secureInputActive: false))
    }

    func testSecureInputStillWinsForEveryTarget() {
        // The pre-existing reason for the clipboard path, unchanged: a secure field cannot receive
        // a synthetic ⌘V at all, whichever app owns it.
        XCTAssertTrue(TextInjector.mustUseClipboard(targetBundleId: nil, secureInputActive: true))
        XCTAssertTrue(TextInjector.mustUseClipboard(targetBundleId: "com.apple.Terminal", secureInputActive: true))
        XCTAssertTrue(TextInjector.mustUseClipboard(targetBundleId: own, secureInputActive: true))
    }

    func testOwnBundleIdIsResolvedFromTheBundleNotHardcodedTwice() {
        XCTAssertEqual(own, "co.miraside.voice")
    }
}

File: windows/Spit.Core.Tests/TextInjectorRoutingTests.cs
namespace Spit.Core.Tests;

/// Port of mac/VoiceTests/TextInjectorRoutingTests.swift (`TextInjector.mustUseClipboard` is
/// `PasteRouting.MustUseClipboard` here). A dictation must never be pasted into Spit itself. The Mac's
/// secure-input cases become elevated-target cases (rules 30, 34): an admin window is the Windows
/// target that cannot receive a synthetic paste.
public sealed class TextInjectorRoutingTests
{
    private static string Own => PasteRouting.OwnBundleId;

    [Fact]
    public void testDictationIntoVoiceItselfGoesToTheClipboard()
    {
        Assert.True(PasteRouting.MustUseClipboard(targetBundleId: Own, targetIsElevated: false));
    }

    [Fact]
    public void testOrdinaryAppsStillGetAPaste()
    {
        foreach (var bundleId in new[] { "chrome.exe", "slack.exe", "notepad.exe", "code.exe" })
        {
            Assert.False(
                PasteRouting.MustUseClipboard(targetBundleId: bundleId, targetIsElevated: false),
                $"{bundleId} must still receive a normal paste");
        }
    }

    [Fact]
    public void testAnUnknownTargetPastesRatherThanBeingTreatedAsVoice()
    {
        // The foreground context can be unknown (no foreground window, access denied on the image name).
        // Reading "I don't know" as "it might be me" would divert ordinary dictations to the clipboard.
        Assert.False(PasteRouting.MustUseClipboard(targetBundleId: null, targetIsElevated: false));
    }

    [Fact]
    public void testSecureInputStillWinsForEveryTarget()
    {
        // Windows' counterpart of secure input: an elevated target cannot receive SendInput at all,
        // whichever app owns it, and the failure is silent.
        Assert.True(PasteRouting.MustUseClipboard(targetBundleId: null, targetIsElevated: true));
        Assert.True(PasteRouting.MustUseClipboard(targetBundleId: "windowsterminal.exe", targetIsElevated: true));
        Assert.True(PasteRouting.MustUseClipboard(targetBundleId: Own, targetIsElevated: true));
    }

    [Fact]
    public void testOwnBundleIdIsResolvedFromTheBundleNotHardcodedTwice()
    {
        // One literal, in the lowercased executable-name form `FrontmostApp.BundleId` carries (rule 31).
        Assert.Equal("spit.exe", Own);
    }
}
